### PR TITLE
Ibimina auth guard

### LINKED ISSUE #97: Guard Managers Route and Add Listing Route
We need to enhance the security of our platform by implementing route guards for the managers' and community listing creation routes. Currently, the managers' routes are accessible without any authentication, which poses a security risk.

Acceptance Criteria:

- Implement route guards to restrict access to the managers' routes.
- Only authenticated managers should be able to access the managers' routes.
- The listing creation route should be accessible to authenticated managers.
- Unauthorized access attempts to the managers' routes should result in an appropriate error response (e.g., 401 Unauthorized or 403 Forbidden).

### PR DESCRIPTION
**Pull request by ibimina**
## What type of PR is this? (check all applicable)

- [x] Feature
- [ ] Refactor
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Other (please specify here: <type here>)
  
  
## Description of what this Pull Request does

- Creates a user context
- Add reducer function to dispatch manager authentication actions
- Display only approved listings
- Refactor login function
- Redirect user after sign in or log in

## Related Tickets & Documents
- Closes #97 
- Related Issues: (enter ticket numbers here)
  
## QA Instructions, Screenshots, Recordings
(In this section, describe how to QA/review your PR. Feel free to use screenshots and/or short videos)
  
## Added/updated tests?
- [ ] Yes
- [ ] Not Yet, and here is why: __explain here__
- [ ] I need help with tests
- [x] No, and here is why: __explain here__

### CODE REVIEW

File: src/context/reducer.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+export const authReducer = ( state, action) => {
+  if (action.type === "SIGN_IN") {
+    return {
+      ...state,
+      user: action.payload,
+    };
+  }
+  if (action.type === "LOG_IN") {
```
Comment (annonymousauthority) [inline src/context/reducer.js:23]: 'SIGN_IN' and 'LOG_IN' are doing same thing

**Review by annonymousauthority — APPROVED** [review]
The change should be effected and behavior of this code monitored, especially around persistence on page reload

File: src/context/reducer.js (hunk 1 above)
Comment (msarit) [reply src/context/reducer.js:25]: @ibimina please explain why we have both sign_in and log_in doing the same thing 🙏🏾

Comment (ibimina) [reply src/context/reducer.js]: I have update the function

File: src/pages/managerDashboard.js [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
+  const  router = useRouter();
 
-const managerDashboard = () => {
-  return (
-    <div>managerDashboard</div>
-  )
-}
+  useEffect(() => {
+    if (state?.user ===undefined || state?.user?.uid === "") {
+    router.push("/");
```
Comment (msarit) [inline src/pages/managerDashboard.js:10]: auth-guard-jump

So @ibimina it appears that your `router.push("/")` implementation here is causing a noticeable "jump" in the user experience (view the GIF above).

I lack the NextJS knowledge so forgive me, but it it possible to render the managerDashboard child components conditionally instead? And then if the logged-in condition is unmet, render an error message to the user, instead of the redirect to index?

@annonymousauthority Please weigh in if you can. 🙏🏾

**Review by msarit — CHANGES_REQUESTED** [review]

File: src/pages/managerDashboard.js (hunk 2 above)
Comment (annonymousauthority) [reply src/pages/managerDashboard.js:10]: She's trying to create a route guard.

@ibimina would you consider making the components group that requires Auth to show loading components/like @msarit said, a would you like to signing screen?

To make it feel natural.

A route guard might be brute for this case.

File: src/pages/login.js [hunk 3]
```diff
@@ -12,36 +13,40 @@ export default function login() {
...
-          />
-        </div> 
-        </section>
-  
-        <section className="mt-10 sm:mx-auto sm:w-full sm:max-w-sm">
-          <LoginFormContainer />
-        </section>
-      </main>
+      <>
+        <Link href="/"> home</Link>
```
Comment (msarit) [inline src/pages/login.js:17]: @ibimina Why did you add these links?

File: src/pages/signup.js [hunk 4]
```diff
@@ -10,28 +11,32 @@ export default function signup() {
...
-            paragraphText={info}
-            headerClassName="font-semibold text-4xl text-left leading-10 text-white"
-            spanClassName="font-semibold block text-4xl leading-15  text-accent-blue"
-            paragraphClassName="text-white text-base text-left self-start tracking-wider mt-2"
-          />
-        </div>
-      </section>
-    </main>
+    <>
+      <Link href="/"> home</Link>
```
Comment (msarit) [inline src/pages/signup.js:15]: @ibimina again, what are these 2 Links for?

File: src/composables/authLogIn.js [hunk 5]
```diff
@@ -1,23 +1,27 @@
...
-    };
-    return { addUser, firebaseError, LogIn };
+  const LogIn = ({ email, password }) => {
+    signInWithEmailAndPassword(firebaseAuth, email, password)
+      .then(async (userCredential) => {
+        // setAddUser(userCredential.user)
+        const docRef = doc(firebaseDb, "MANAGERS", userCredential?.user?.uid);
+        const docSnap = await getDoc(docRef);
+        dispatch({ type: "SIGN_IN", payload: docSnap.data() });
+        //  Handle successful login
```
Comment (msarit) [inline src/composables/authLogIn.js:18]: @ibimina Do we actually handle the successful login? When I login to the local app, my login credentials print to console but I am not redirected to `managerDashboard`

**Comment by msarit** [comment]
@ibimina Please my dear this PR needs your attention

File: src/composables/authLogIn.js (hunk 5 above)
Comment (ibimina) [reply src/composables/authLogIn.js:18]: After a successful login the user should be redirected to the managerDashboard page

Comment (ibimina) [reply src/composables/authLogIn.js]: @msarit I refactored the login function, The login function was not working because it was not hooked up with firebase

**Review by jjleigh — APPROVED** [review]

**Review by msarit — APPROVED** [review]
